File: Swift/프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.swift
import Foundation

func solution(_ enroll:[String], _ referral:[String], _ seller:[String], _ amount:[Int]) -> [Int] {
    var parent: [String: String] = [:]
    for i in 0..<enroll.count {
        parent[enroll[i]] = referral[i]
    }
    
    var total: [String: Int] = [:]
    
    for i in 0..<seller.count {
        var name = seller[i]
        var money = amount[i] * 100
        
        while money > 0 && name != "-" {
            total[name] = (total[name] ?? 0) + money - (money / 10)
            name = parent[name] ?? ""
            
            money /= 10
        }
    }
    
    var answer: [Int] = .init(repeating: 0, count: enroll.count)
    for i in 0..<enroll.count {
        answer[i] = total[enroll[i]] ?? 0
    }
    return answer
}

File: Java/프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.java
import java.util.HashMap;

class Solution {
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        int enrollLength = enroll.length;
        //추천인 link 구조 만들기
        HashMap<String, String> parentMap = new HashMap<>();
        for (int i = 0; i<enrollLength; i++) {
            parentMap.put(enroll[i], referral[i]);
        }
        
        //수익금 종합
        HashMap<String, Integer> sellingResult = new HashMap<>();
        for (int i = 0; i<seller.length; i++) {
            String person = seller[i];
            Integer amountMoney = amount[i] * 100;
            
            while (amountMoney > 0 && !person.equals("-")) {
                
                sellingResult.put(person, sellingResult.getOrDefault(person, 0) + amountMoney - (amountMoney/10));
                person = parentMap.getOrDefault(person, "-");
                amountMoney = amountMoney / 10;
                }
        }
        
        int[] answer = new int[enrollLength];
        for (int i = 0; i<enrollLength; i++) {
            String person = enroll[i];
            answer[i] = sellingResult.getOrDefault(person, 0);
        }
        
        return answer;
    }
}